### ISSUE TITLE
label metrics by route

### ISSUE THREAD

**Issue by m-ronchi** [Ptrskay3/axum-prometheus#4]
Hi,
is it possible to label the metrics by configured route instead of endpoint?

e.g. if I have a route with path parameters (e.g. `/foo/:bar/:baz`) I want to get one set of metrics (or optionally one per allowed value of :bar and :baz by adding additional labels) instead of one per actual endpoint hit by anyone

currently there is a possible problem with exploding metrics if path params have high cardinality, either intentionally or someone trying to break our APIs

**Comment by Ptrskay3** [comment]
That is almost good, but sadly won't work with [nested routers](https://docs.rs/axum/latest/axum/extract/struct.MatchedPath.html#matched-path-in-nested-routers). This means if we do the same thing as in the axum example, we would still get a unique entry in Prometheus per unique path for nested routers.

**Comment by Ptrskay3** [comment]
I started to experiment in #5 with the solution I mentioned above. What do you think about it?

**Comment by m-ronchi** [comment]
it should be fine as a workaround (or low-level api), but it implies to duplicate all route paths into the metric builder.

ideally the layer can get all routes from axum and use them internally, and the user can configure if it wants the endpoint or route (and if the route doesn't match, if it fallbacks to the endpoint or a constant label)

ps: with this API, can I tell it to group all unknown endpoints (i.e. when nothing matches) into one? (or to ignore anything that doesn't match a group?)

**Comment by m-ronchi** [comment]
> As far as I can see now, we can't get the inner routing information from axum other than `MatchedPath` or `request.uri()`. Even if we could, I have no idea how to even hook into the existing router to accept additional parameters (assuming you mean to be able to put an additional parameter into the original axum router that affects metrics) - that's beyond the scope of what a middleware should be able to do. Again, probably it is possible to achieve that with macros, but I'm not planning to go into that direction with this crate.
> 

no I mean in the metrics builder. something like this:
```rust
let prometheus_layer = PrometheusMetricLayer::builder()
    .with_endpoint_label_type(Type::Endpoint) // or ::Route, or ::RouteWithFallback("unknown"), etc...
```
and the middleware internally switches between using `uri()` or `MatchedPath` or something else


> You can filter for the status label != 404 though

I solved this by adding the layer using `route_layer` instead of `layer` so it only attaches to actual routes

**Comment by Ptrskay3** [comment]
I guess we could add `with_endpoint_label_type` to the builder, but the fallback variant won't work because of what I mentioned above. 

One thing that feels unfortunate that it's sort of ambiguous to use `MatchedPath` - people will get confused when they have nested routers and suddenly it starts falling back to `uri()`. That being said, probably with careful defaults and good documentation we could eventually support this. I'll try to experiment.

**Comment by Ptrskay3** [comment]
Just released 0.3, which should solve this issue, see #5 and #8. I hope it suits your needs.
